ethers-contract/src/multicall.rs

```diff
@@ -0,0 +1,204 @@
+        let calls: Vec<Call> = calls
+            .into_iter()
+            .filter_map(|call| match (call.tx.to, call.tx.data) {
+                (Some(NameOrAddress::Address(target)), Some(data)) => Some(Call {
+                    target,
+                    data: data.0,
+                    function: call.function,
+                }),
+                _ => None,
+            })
+            .collect();
+
+        self.calls.extend(calls);
+        self
+    }
+}
+
+impl<P, S> Multicall<P, S>
+where
+    P: JsonRpcClient,
+    S: Signer,
+{
+    pub async fn call<D: Detokenize>(&self) -> Result<(U256, D), ContractError> {
+        // 1. Fetch chain id and the corresponding address of Multicall contract
+        // preference is given to Multicall contract's address if provided
+        // otherwise check the address book for the client's chain ID
+        let address = match self.address {
+            Some(address) => address,
+            None => {
+                let chain_id = self.client.get_chainid().await?;
+                match ADDRESS_BOOK.get(&chain_id) {
+                    Some(address) => address.clone(),
+                    None => return Err(ContractError::ConstructorError),
+                }
+            }
+        };
+
+        // 2. Instantiate the multicall contract
+        let multicall = MulticallContract::new(address, Arc::clone(&self.client));
+
+        // 3. Map the Multicall struct into appropriate types for `aggregate` function
+        let calls: Vec<(Address, Vec<u8>)> = self
+            .calls
+            .clone()
+            .into_iter()
+            .map(|call| (call.target, call.data))
+            .collect();
+
+        // 4. Call the `aggregate` function and get return data
+        let contract_call = multicall.aggregate(calls);
+        let contract_call = {
+            if let Some(block) = self.block {
+                contract_call.block(block)
+            } else {
+                contract_call
+            }
+        };
+
+        let (block_number, vec_bytes) = contract_call.call().await?;
+
+        // 5. Decode return data into ABI tokens
+        let tokens: Vec<Token> = self
+            .calls
+            .clone()
+            .into_iter()
+            .zip(vec_bytes.into_iter())
+            .map(|(call, bytes)| {
+                let tokens: Vec<Token> = call.function.decode_output(&bytes).unwrap();
+
+                // NOTE: post processing
+                match tokens.len() {
+                    0 => Token::Tuple(vec![]),
+                    1 => tokens[0].clone(),
+                    _ => Token::Tuple(tokens),
+                }
+            })
+            .collect();
+
+        // 6. Form tokens that represent tuples
+        let tokens = vec![Token::Tuple(tokens)];
+
+        // 7. Detokenize from the tokens into the provided tuple D
+        let data = D::from_tokens(tokens)?;
+
+        Ok((block_number, data))
```

Comment:
I think we can drop the block_number return argument, it doesn't feel that useful, and I'm a bit confused by why they added it in the original multicall contract